test/test/lookup_as.cpp

```diff
@@ -0,0 +1,82 @@
+#include "pch.h"
+
+using namespace winrt;
+using namespace Windows::Foundation;
+using namespace Windows::Foundation::Collections;
+
+namespace
+{
+    template <typename M>
+    void test(M const& m)
+    {
+        // Good LookupAs
+
+        REQUIRE(m.LookupAs<Uri>(L"uri").Domain() == L"kennykerr.ca");
+        REQUIRE(m.LookupAs<IStringable>(L"uri").ToString() == L"http://kennykerr.ca/");
+        REQUIRE(m.LookupAs<AsyncStatus>(L"enum") == AsyncStatus::Completed);
+        REQUIRE(m.LookupAs<AsyncStatus>(L"underlying_type") == AsyncStatus::Error);
+        REQUIRE(m.LookupAs<guid>(L"guid") == guid_of<Uri>());
+        REQUIRE(m.LookupAs<GUID>(L"guid") == static_cast<GUID>(guid_of<Uri>()));
+        REQUIRE(m.LookupAs<hstring>(L"hstring") == L"Test");
+
+        // LookupAs edge cases
+
+        REQUIRE_THROWS_AS(m.LookupAs<bool>(L"INVALID"), hresult_out_of_bounds);
+        REQUIRE(m.LookupAs<Uri>(L"nullptr") == nullptr);
+        REQUIRE_THROWS_AS(m.LookupAs<bool>(L"nullptr"), hresult_no_interface);
+
+        // Bad LookupAs
+
+        REQUIRE_THROWS_AS(m.LookupAs<bool>(L"uri"), hresult_no_interface);
+        REQUIRE_THROWS_AS(m.LookupAs<IClosable>(L"uri"), hresult_no_interface);
+        REQUIRE_THROWS_AS(m.LookupAs<AsyncStatus>(L"uri"), hresult_no_interface);
+        REQUIRE_THROWS_AS(m.LookupAs<guid>(L"uri"), hresult_no_interface);
+        REQUIRE_THROWS_AS(m.LookupAs<GUID>(L"uri"), hresult_no_interface);
+
+        // Good TryLookupAs (these tests don't need `.value()` for comparison but it
+        // helps to assert that those variants return std::optional)
+
+        REQUIRE(m.TryLookupAs<Uri>(L"uri").Domain() == L"kennykerr.ca");
+        REQUIRE(m.TryLookupAs<IStringable>(L"uri").ToString() == L"http://kennykerr.ca/");
+        REQUIRE(m.TryLookupAs<AsyncStatus>(L"enum").value() == AsyncStatus::Completed);
+        REQUIRE(m.TryLookupAs<AsyncStatus>(L"underlying_type").value() == AsyncStatus::Error);
+        REQUIRE(m.TryLookupAs<guid>(L"guid").value() == guid_of<Uri>());
+        REQUIRE(m.TryLookupAs<GUID>(L"guid").value() == static_cast<GUID>(guid_of<Uri>()));
+        REQUIRE(m.TryLookupAs<hstring>(L"hstring").value() == L"Test");
+
+        // TryLookupAs edge cases
+
+        REQUIRE_FALSE(m.TryLookupAs<bool>(L"INVALID").has_value());
```

Comment:
Does "nullptr" here represent the invalid value? Or should there be a test for `m.TryLookupAs<Uri>(L"INVALID")` here?

Reply:
`TryLookupAs<bool>` will return a `std::optional<bool>`, and we are verifying that it returns a valueless object.

In the second test, `TryLookupAs<Uri>` will return a `Uri`, and we are using the `Uri == nullptr` test to see if the `Uri` object is empty.

Reply:
How about a test that validates `TryLookupAs<IInspectableDerivedThing>(L"VALUE THAT IS NOT IN THE MAP")`?